File: LISTAS ENLAZADAS/listasenlazadas.py
import os

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
    
    def clear_screen(self):
        os.system('cls' if os.name == 'nt' else 'clear')
    
    def beginset(self):
        self.clear_screen()
        print("=== INSERTAR AL INICIO ===")
        item = int(input("Ingrese el valor del nodo: "))
        new_node = Node(item)
        new_node.next = self.head
        self.head = new_node
        print("Nodo insertado al inicio correctamente")
    
    def lastinsert(self):
        self.clear_screen()
        print("=== INSERTAR AL FINAL ===")
        item = int(input("Ingrese el valor del nodo: "))
        new_node = Node(item)
        
        if self.head is None:
            self.head = new_node
            print("Nodo insertado al final (lista estaba vacia)")
        else:
            temp = self.head
            while temp.next is not None:
                temp = temp.next
            temp.next = new_node
            print("Nodo insertado al final correctamente")
    
    def randominsert(self):
        self.clear_screen()
        print("=== INSERTAR EN POSICION ESPECIFICA ===")
        item = int(input("Ingrese el valor del elemento: "))
        loc = int(input("Ingrese la posicion despues de la cual desea insertar: "))
        
        new_node = Node(item)
        
        if self.head is None:
            print("La lista esta vacia. Insertando al inicio...")
            self.head = new_node
        else:
            temp = self.head
            for i in range(loc):
                if temp is None:
                    print("No se puede insertar en esa posicion")
                    return
                temp = temp.next
            
            if temp is None:
                print("No se puede insertar en esa posicion")
                return
            
            new_node.next = temp.next
            temp.next = new_node
            print(f"Nodo insertado correctamente en la posicion {loc + 1}")
    
    def begin_delete(self):
        self.clear_screen()
        print("=== ELIMINAR DEL INICIO ===")
        if self.head is None:
            print("La lista esta vacia")
        else:
            temp = self.head
            self.head = self.head.next
            temp = None
            print("Nodo eliminado del inicio correctamente")
    
    def last_delete(self):
        self.clear_screen()
        print("=== ELIMINAR DEL FINAL ===")
        if self.head is None:
            print("La lista esta vacia")
        elif self.head.next is None:
            self.head = None
            print("Unico nodo eliminado")
        else:
            temp = self.head
            while temp.next is not None:
                prev = temp
                temp = temp.next
            prev.next = None
            temp = None
            print("Nodo eliminado del final correctamente")
    
    def random_delete(self):
        self.clear_screen()
        print("=== ELIMINAR NODO EN POSICION ESPECIFICA ===")
        loc = int(input("Ingrese la posicion del nodo que desea eliminar: "))
        
        if self.head is None:
            print("La lista esta vacia")
            return
        
        if loc == 0:
            temp = self.head
            self.head = self.head.next
            temp = None
            print(f"Nodo eliminado de la posicion {loc}")
            return
        
        temp = self.head
        for i in range(loc):
            prev = temp
            temp = temp.next
            if temp is None:
                print("No se puede eliminar - posicion fuera de rango")
                return
        
        prev.next = temp.next
        temp = None
        print(f"Nodo eliminado de la posicion {loc}")
    
    def search(self):
        self.clear_screen()
        print("=== BUSCAR ELEMENTO ===")
        if self.head is None:
            print("Lista vacia")
            return
        
        item = int(input("Ingrese el elemento que desea buscar: "))
        temp = self.head
        i = 0
        found = False
        
        while temp is not None:
            if temp.data == item:
                print(f"Elemento encontrado en la posicion {i + 1}")
                found = True
                break
            i += 1
            temp = temp.next
        
        if not found:
            print("Elemento no encontrado")
    
    def display(self):
        self.clear_screen()
        print("=== MOSTRAR LISTA ===")
        if self.head is None:
            print("La lista esta vacia")
        else:
            temp = self.head
            print("Elementos de la lista: ", end="")
            while temp is not None:
                print(temp.data, end=" ")
                temp = temp.next
            print()
    
    def menu(self):
        while True:
            self.clear_screen()
            print("\n\n**Menu Principal")
            print("\nElija una opcion del siguiente menu:")
            print("1. Insertar al inicio")
            print("2. Insertar al final")
            print("3. Insertar en posicion aleatoria")
            print("4. Eliminar del inicio")
            print("5. Eliminar del final")
            print("6. Eliminar nodo despues de una posicion")
            print("7. Buscar elemento")
            print("8. Mostrar lista")
            print("9. Salir")
            
            try:
                choice = int(input("\nIngrese su eleccion: "))
                
                if choice == 1:
                    self.beginset()
                elif choice == 2:
                    self.lastinsert()
                elif choice == 3:
                    self.randominsert()
                elif choice == 4:
                    self.begin_delete()
                elif choice == 5:
                    self.last_delete()
                elif choice == 6:
                    self.random_delete()
                elif choice == 7:
                    self.search()
                elif choice == 8:
                    self.display()
                elif choice == 9:
                    self.clear_screen()
                    print("Saliendo del programa...")
                    break
                else:
                    self.clear_screen()
                    print("Por favor ingrese una opcion valida...")
                
                if choice != 9:
                    input("\n\nPresione Enter para continuar...")
                    
            except ValueError:
                self.clear_screen()
                print("Por favor ingrese un numero valido...")
                input("\n\nPresione Enter para continuar...")

# Ejecutar el programa
if __name__ == "__main__":
    lista = LinkedList()
    lista.menu()

File: LISTAS ENLAZADAS/listasenlazadas.cpp
#include <iostream>  // Para operaciones de entrada/salida (cout, cin)
#include <cstdlib>  // Para funciones del sistema como exit()
#ifdef _WIN32
    #include <windows.h>  // Para system("cls") en Windows
#else
    #include <unistd.h>  // Para system("clear") en Linux/Mac
#endif
using namespace std;  // Uso del espacio de nombres estándar

// Definición de la estructura del nodo para la lista enlazada
struct node {
    int data;    // Almacena el valor numérico del nodo
    struct node *next; // Puntero que apunta al siguiente nodo en la lista
};

// Variable global que apunta al primer nodo de la lista
// Se inicializa automáticamente a NULL al ser una variable global
struct node *head;

// Función para limpiar la pantalla
void clearScreen() {
    #ifdef _WIN32
        system("cls");
    #else
        system("clear");
    #endif
}

// Declaración de funciones para operaciones de la lista enlazada
void beginset();   // Inserta un nuevo nodo al principio de la lista
void lastinsert();   // Inserta un nuevo nodo al final de la lista
void randominsert();   // Inserta un nuevo nodo después de una posición específica
void begin_delete();   // Elimina el primer nodo de la lista
void last_delete();   // Elimina el último nodo de la lista
void random_delete();   // Elimina un nodo después de una posición específica
void display();   // Muestra todos los elementos de la lista
void search();   // Busca un elemento específico en la lista y muestra su posición

// Función principal
int main() {
    int choice = 0;  // variable para la elección del usuario
    while (choice != 9) {  // ciclo hasta que el usuario elija salir
        clearScreen();  // Limpiar pantalla al inicio de cada ciclo
        cout << "\n\n**Menu Principal\n";  // Título del menú
        cout << "\nElija una opcion del siguiente menu:\n";  // Instrucción para elegir una opción
        // Opciones del menú
        cout << "1.Insertar al inicio\n2.Insertar al final\n3.Insertar en posicion aleatoria\n4.Eliminar del inicio\n5.Eliminar del final\n6.Eliminar nodo despues de una posicion\n7.Buscar elemento\n8.Mostrar lista\n9.Salir\n";
        cout << "\nIngrese su eleccion: ";  // Solicita la opción del usuario
        cin >> choice;  // Lee la opción del usuario

        switch (choice) {  // Evalúa la opción seleccionada
            case 1:
                clearScreen();
                beginset();  // Llama a la función para insertar al principio
                break;
            case 2:
                clearScreen();
                lastinsert();  // Llama a la función para insertar al final
                break;
            case 3:
                clearScreen();
                randominsert();  // Llama a la función para insertar en una posición específica
                break;
            case 4:
                clearScreen();
                begin_delete();  // Llama a la función para eliminar el primer nodo
                break;
            case 5:
                clearScreen();
                last_delete();  // Llama a la función para eliminar el último nodo
                break;
            case 6:
                clearScreen();
                random_delete();  // Llama a la función para eliminar un nodo en una posición específica
                break;
            case 7:
                clearScreen();
                search();  // Llama a la función para buscar un elemento en la lista
                break;
            case 8:
                clearScreen();
                display();  // Llama a la función para mostrar todos los elementos de la lista
                break;
            case 9:
                clearScreen();
                cout << "Saliendo del programa...\n";
                exit(0);  // Sale del programa
                break;
            default:
                clearScreen();
                cout << "Por favor ingrese una opcion valida...";  // Maneja opciones inválidas
        }

        // Pausa antes de volver al menú (solo si no es salir)
        if (choice != 9) {
            cout << "\n\nPresione Enter para continuar...";
            cin.ignore();
            cin.get();
        }
    }
    return 0;
}

void beginset() {
    struct node *ptr;
    int item;
    ptr = new node;
    if (ptr == NULL) {
        cout << "OVERFLOW - No hay memoria disponible\n";
    } else {
        cout << "=== INSERTAR AL INICIO ===\n";
        cout << "Ingrese el valor del nodo: ";
        cin >> item;
        ptr->data = item;
        ptr->next = head;
        head = ptr;
        cout << "Nodo insertado al inicio correctamente\n";
    }
}

void lastinsert() {
    struct node *ptr, *temp;
    int item;
    ptr = new node;
    if (ptr == NULL) {
        cout << "OVERFLOW - No hay memoria disponible\n";
    } else {
        cout << "=== INSERTAR AL FINAL ===\n";
        cout << "Ingrese el valor del nodo: ";
        cin >> item;
        ptr->data = item;
        if (head == NULL) {
            ptr->next = NULL;
            head = ptr;
            cout << "Nodo insertado al final (lista estaba vacia)\n";
        } else {
            temp = head;
            while (temp->next != NULL) {
                temp = temp->next;
            }
            temp->next = ptr;
            ptr->next = NULL;
            cout << "Nodo insertado al final correctamente\n";
        }
    }
}

void randominsert() {
    int i, loc, item;
    struct node *ptr, *temp;
    ptr = new node;
    if (ptr == NULL) {
        cout << "OVERFLOW - No hay memoria disponible\n";
    } else {
        cout << "=== INSERTAR EN POSICION ESPECIFICA ===\n";
        cout << "Ingrese el valor del elemento: ";
        cin >> item;
        ptr->data = item;
        cout << "Ingrese la posicion despues de la cual desea insertar: ";
        cin >> loc;
        temp = head;

        if (head == NULL) {
            cout << "La lista esta vacia. Insertando al inicio...\n";
            ptr->next = NULL;
            head = ptr;
        } else {
            for (i = 0; i < loc; i++) {
                if (temp == NULL) {
                    cout << "No se puede insertar en esa posicion\n";
                    delete ptr;
                    return;
                }
                temp = temp->next;
            }
            ptr->next = temp->next;
            temp->next = ptr;
        }
        cout << "Nodo insertado correctamente en la posicion " << loc + 1 << endl;
    }
}

void begin_delete() {
    struct node *ptr;
    cout << "=== ELIMINAR DEL INICIO ===\n";
    if (head == NULL) {
        cout << "La lista esta vacia\n";
    } else {
        ptr = head;
        head = ptr->next;
        delete ptr;
        cout << "Nodo eliminado del inicio correctamente\n";
    }
}

void last_delete() {
    struct node *ptr, *ptr1;
    cout << "=== ELIMINAR DEL FINAL ===\n";
    if (head == NULL) {
        cout << "La lista esta vacia\n";
    } else if (head->next == NULL) {
        delete head;
        head = NULL;
        cout << "Unico nodo eliminado\n";
    } else {
        ptr = head;
        while (ptr->next != NULL) {
            ptr1 = ptr;
            ptr = ptr->next;
        }
        ptr1->next = NULL;
        delete ptr;
        cout << "Nodo eliminado del final correctamente\n";
    }
}

void random_delete() {
    struct node *ptr, *ptr1;
    int loc, i;
    cout << "=== ELIMINAR NODO EN POSICION ESPECIFICA ===\n";
    cout << "Ingrese la posicion del nodo que desea eliminar: ";
    cin >> loc;

    if (head == NULL) {
        cout << "La lista esta vacia\n";
        return;
    }

    ptr = head;
    if (loc == 0) {
        head = ptr->next;
        delete ptr;
        cout << "Nodo eliminado de la posicion " << loc << endl;
        return;
    }

    for (i = 0; i < loc; i++) {
        ptr1 = ptr;
        ptr = ptr->next;
        if (ptr == NULL) {
            cout << "No se puede eliminar - posicion fuera de rango\n";
            return;
        }
    }
    ptr1->next = ptr->next;
    delete ptr;
    cout << "Nodo eliminado de la posicion " << loc << endl;
}

void search() {
    struct node *ptr;
    int item, i = 0, flag = 1;
    cout << "=== BUSCAR ELEMENTO ===\n";
    ptr = head;
    if (ptr == NULL) {
        cout << "Lista vacia\n";
    } else {
        cout << "Ingrese el elemento que desea buscar: ";
        cin >> item;
        while (ptr != NULL) {
            if (ptr->data == item) {
                cout << "Elemento encontrado en la posicion " << i + 1 << endl;
                flag = 0;
                break;
            }
            i++;
            ptr = ptr->next;
        }
        if (flag == 1) {
            cout << "Elemento no encontrado\n";
        }
    }
}

void display() {
    struct node *ptr;
    cout << "=== MOSTRAR LISTA ===\n";
    ptr = head;
    if (ptr == NULL) {
        cout << "La lista esta vacia";
    } else {
        cout << "Elementos de la lista: ";
        while (ptr != NULL) {
            cout << ptr->data << " ";
            ptr = ptr->next;
        }
        cout << endl;
    }
}
